### PR TITLE
Rewrote leftAlignIndels code -- it now always works, even for multiple indels

### LINKED ISSUE #150: add tests for LeftAlignIndels
there were none in GATK, need to make some in hellbender. include cram tests

### LINKED ISSUE #4919: leftAlignIndel only works for reads with one indel
`AlignmentUtils.leftAlignIndel` doesn't modify the CIGAR string if it contains more than one indel.

### PR DESCRIPTION
**Pull request by davidbenjamin**
@kachulis I hope you find this entertaining.

Closes #150.

Closes #4919. 

Also fixes a bug we didn't know about where `LeftAlignAndTrimVariants` didn't do anything when `--dont-trim` was set.

Also fixes a bug where `LeftAlignAndTrimVariants` couldn't left align an indel past a SNP.

Also fixes a bug where `LeftAlignAndTrimVariants` didn't left align multiallelic variants when splitting was turned off.

Probably fixes a few other bugs since the old code failed a bunch of my unit tests.

### CODE REVIEW

File: src/main/java/org/broadinstitute/hellbender/utils/read/AlignmentUtils.java [hunk 1]
```diff
@@ -763,178 +766,166 @@ private static boolean needsConsolidation(final Cigar c) {
...
-        if ( doNotThrowExceptionForMultipleIndels )
-            return cigar;
+        // we need reference bases from the start of the read to the rightmost indel
+        final int lastIndel = IntStream.range(0, cigar.numCigarElements()).filter(n -> cigar.getCigarElement(n).getOperator().isIndel()).max().getAsInt();
+        final int necessaryRefLength = readStart + cigar.getCigarElements().stream().limit(lastIndel + 1).mapToInt(e -> lengthOnReference(e)).sum();
+        Utils.validateArg(necessaryRefLength <= ref.length, "read goes past end of reference");
+
+        // at this point, we are one base past the end of the read.  Now we traverse the cigar from right to left
+        final List<CigarElement> resultRightToLeft = new ArrayList<>();
+        final int refLength = cigar.getCigarElements().stream().mapToInt(e -> lengthOnReference(e)).sum();
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/read/AlignmentUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/read/AlignmentUtils.java [hunk 2]
```diff
@@ -763,178 +766,166 @@ private static boolean needsConsolidation(final Cigar c) {
...
+        Utils.validateArg(sequences.size() == bounds.size(), "Must have one initial allele range per sequence");
+        bounds.forEach(bound -> Utils.validateArg(maxShift <= bound.getStart(), "maxShift goes past the start of a sequence"));
+
+        int startShift = 0;
+        int endShift = 0;
+
+        // consume any redundant shared bases at the end of the alleles
+        int minSize = bounds.stream().mapToInt(IndexRange::size).min().getAsInt();
+        while (minSize > 0) {
+            if (lastBaseOnRightIsSame(sequences, bounds)) {
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/read/AlignmentUtils.java]: True.  done

File: src/main/java/org/broadinstitute/hellbender/utils/read/AlignmentUtils.java [hunk 3]
```diff
@@ -763,178 +766,166 @@ private static boolean needsConsolidation(final Cigar c) {
...
-
-    /**
-     * See the documentation for AlignmentUtils.leftAlignIndel() for more details.
+     *  If the given allele ranges are not parsimonious, for example [3,4) and [3,5) in the above example to include the common T in both alleles,
+     *  the resulting ranges will be shifted by different amounts.  In this case, the shifts are 2 bases in the front and 3 at the end.
      *
-     * This flavor of the left alignment works if and only if the alignment has one - and only one - indel.
-     * An exception is thrown if there are no indels or more than 1 indel in the alignment.
+     *  Note that we use the convention that the ref allele in the case of an alt insertion, or the alt allele in case of a deletion, is represented
+     *  by [n,n) where n is the last aligned coordinate before the indel.  This makes sense when you think in terms of alignment CIGARs:
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/read/AlignmentUtils.java]: done

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/LeftAlignAndTrimVariants.java [hunk 4]
```diff
@@ -175,39 +176,21 @@
...
-    int longestSkippedVariant;
-    private int furthestEndOfEarlierVariant;
-    private String currentContig;
 
-    /**
-     * Set up the VCF writer, samples
-     */
+    // it is possible that left-alignment changes the order of variants.  For example, an indel in an STR may left align past a
+    // SNP in the same STR that doesn't move.  Thus we can only emit variants that are sufficiently far behind the current one.
+    PriorityQueue<VariantContext> pendingVariants;
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/LeftAlignAndTrimVariants.java]: Thanks for the catch and the offline explanation.  Done.

File: src/main/java/org/broadinstitute/hellbender/tools/LeftAlignIndels.java [hunk 5]
```diff
@@ -66,25 +64,19 @@ public boolean requiresReference() {
...
 
-        // move existing indels (for 1 indel reads only) to leftmost position within identical sequence
-        int numBlocks = AlignmentUtils.getNumAlignmentBlocks(read);
-        if ( numBlocks == 2 ) {
-            // We checked in onTraversalStart() that a reference is present, so ref.get() is safe
-            Cigar newCigar = AlignmentUtils.leftAlignIndel(CigarUtils.trimReadToUnclippedBases(read.getCigar()), ref.getBases(), read.getBases(), 0, 0, true);
-            newCigar = CigarUtils.reclipCigar(newCigar, read);
-            read.setCigar(newCigar);
-        }
+        Cigar newCigar = AlignmentUtils.leftAlignIndels(read.getCigar(), ref.getBases(), read.getBases(), 0);
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/LeftAlignIndels.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/read/AlignmentUtils.java [hunk 6]
```diff
@@ -763,178 +766,166 @@ private static boolean needsConsolidation(final Cigar c) {
...
+                resultRightToLeft.add(element);
+                refIndelRange.shiftLeft(lengthOnReference(element));
+                readIndelRange.shiftLeft(lengthOnRead(element));
+            } else {    // there's an indel that we need to left-align
+                // we can left-align into match cigar elements but not into clips
+                final int maxShift = element.getOperator().isAlignment() ? element.getLength() : 0;
+                final Pair<Integer, Integer> shifts = leftAlignAlleles(Arrays.asList(ref, read), Arrays.asList(refIndelRange, readIndelRange), maxShift);
+
+                // account for new match alignments on the right due to left-alignment
+                resultRightToLeft.add(new CigarElement(shifts.getRight(), CigarOperator.MATCH_OR_MISMATCH));
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/read/AlignmentUtils.java]: It's different here because below we're re-emitting a shorter version of a cigar element that already existed.  Here we're creating something new resulting from the left-alignment.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ReadPosRankSumTest.java [hunk 7]
```diff
@@ -69,6 +70,9 @@ public static OptionalDouble getReadPosition(final GATKRead read, final int refL
         final int leadingHardClips = firstElement.getOperator() == CigarOperator.HARD_CLIP ? firstElement.getLength() : 0;
         final int trailingHardClips = lastElement.getOperator() == CigarOperator.HARD_CLIP ? lastElement.getLength() : 0;
 
+        if (offset >= cigar.getReadLength()) {
+            return OptionalDouble.empty();
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/ReadPosRankSumTest.java]: It's intentional.  A test was failing because correct left-alignment exposed an edge case.  I'm about to issue a PR related to `ReadPosRankSumTest` that will re-do this code.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/LeftAlignAndTrimVariants.java [hunk 8]
```diff
@@ -352,127 +289,56 @@ protected int trimAlign(final VariantContext vc, final ReferenceContext ref) {
...
-            // left align the CIGAR
-            Cigar newCigar = AlignmentUtils.leftAlignIndel(originalCigar, refSeq, originalIndel, 0, 0, leftmostAllowedAlignment, true);
-            if (newCigar.equals(originalCigar)) {
+            final List<IndexRange> alleleRanges = vc.getAlleles().stream()
+                    .map(a -> new IndexRange(variantOffsetInRef + 1, variantOffsetInRef + a.length()))
+                    .collect(Collectors.toList());
+
+            // note that this also shifts the index ranges as a side effect, so below they can be used to output allele bases
+            // since trimming is performed elsewhere, we care only about the start shift
+            final int shift = AlignmentUtils.leftAlignAlleles(sequences, alleleRanges, variantOffsetInRef).getLeft();
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/LeftAlignAndTrimVariants.java]: done

File: src/main/java/org/broadinstitute/hellbender/utils/read/AlignmentUtils.java [hunk 9]
```diff
@@ -763,178 +766,166 @@ private static boolean needsConsolidation(final Cigar c) {
...
-
-        int indexOfIndel = -1;
-        for (int i = 0; i < cigar.numCigarElements(); i++) {
-            CigarElement ce = cigar.getCigarElement(i);
-            if (ce.getOperator() == CigarOperator.D || ce.getOperator() == CigarOperator.I) {
-                // if there is more than 1 indel, exception out
-                if (indexOfIndel != -1)
-                    throw new IllegalArgumentException("attempting to left align a CIGAR that has more than 1 indel in its alignment");
-                indexOfIndel = i;
+    public static Pair<Integer, Integer> leftAlignAlleles(final List<byte[]> sequences, final List<IndexRange> bounds, final int maxShift) {
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/utils/read/AlignmentUtils.java]: done

**Comment by davidbenjamin** [comment]
@kachulis back to you

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/LeftAlignAndTrimVariants.java [hunk 10]
```diff
@@ -352,127 +273,56 @@ protected int trimAlign(final VariantContext vc, final ReferenceContext ref) {
...
-        if (!vc.isSimpleIndel()) {
+    @VisibleForTesting
+    static VariantContext leftAlignAndTrim(final VariantContext vc, final ReferenceContext ref, final int maxLeadingBases, final boolean trim) {
+        if (!vc.isIndel() || maxLeadingBases <= 0) {
             return vc;
         }
 
-        // get the indel length
-        final int indelLength = Math.abs(vc.getIndelLengths().get(0));
+        for(int leadingBases = Math.min(maxLeadingBases, 10); leadingBases <= maxLeadingBases; leadingBases *= 2) {
```
Comment (kachulis) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/LeftAlignAndTrimVariants.java]: I believe there is a bug here (quite possibly originating prior to this PR).  If `maxLeadingBases` is greater than 10, but not a multiple of 10, there can be problems.  Suppose `maxLeadingBases=15`, and the indel should be left aligned that full amount.  `leadingBases` will initialize to 10, and the indel will left align the full 10 bases.  `leadingBases` will then double to 20, the for loop condition will be false, and the variant won't be shifted left at all.  I think this is easy to fix, just needs a slight adjustment to the control flow logic.

Also would be nice to have a test of this type of case, to ensure it is handled correctly.

Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/LeftAlignAndTrimVariants.java]: done

**Comment by davidbenjamin** [comment]
@kachulis Back to you.  I fixed the bug and wrote a test for it.  I also put in a commit replacing some trimming code with the new normalization method.

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/LeftAlignAndTrimVariants.java [hunk 11]
```diff
@@ -279,7 +279,8 @@ static VariantContext leftAlignAndTrim(final VariantContext vc, final ReferenceC
             return vc;
         }
 
-        for(int leadingBases = Math.min(maxLeadingBases, 10); leadingBases <= maxLeadingBases; leadingBases *= 2) {
+
+        for(int leadingBases = Math.min(maxLeadingBases, 10); leadingBases <= maxLeadingBases; leadingBases = Math.min(2*leadingBases, maxLeadingBases)) {
             final int refStart = Math.max(vc.getStart() - Math.min(leadingBases, maxLeadingBases), 1);
```
[the remark this answers is not in the document]
Comment (davidbenjamin) [reply src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/LeftAlignAndTrimVariants.java]: I agree.
